net/pubsub.go

```diff
@@ -0,0 +1,165 @@
+	return &TopicManager{
+		ctx:    ctx,
+		host:   host,
+		ps:     ps,
+		handle: handler,
+		m:      make(map[thread.ID]*topic),
+	}
+}
+
+// Add a new thread topic. This may be called repeatedly for the same thread.
+func (m *TopicManager) Add(id thread.ID) error {
+	m.Lock()
+	defer m.Unlock()
+	if _, ok := m.m[id]; ok {
+		return nil
+	}
+
+	pt, err := m.ps.Join(id.String())
+	if err != nil {
+		return err
+	}
+	h, err := pt.EventHandler()
+	if err != nil {
+		return err
+	}
+	if err = m.ps.RegisterTopicValidator(id.String(), m.topicValidator); err != nil {
+		return err
+	}
+
+	topic := &topic{t: pt, h: h}
+	m.m[id] = topic
+	go m.watch(m.ctx, id, topic)
+	go m.subscribe(m.ctx, id, topic)
+	return nil
+}
+
+// Remove a thread topic. This may be called repeatedly for the same thread.
+func (m *TopicManager) Remove(id thread.ID) error {
+	m.Lock()
+	defer m.Unlock()
+	topic, ok := m.m[id]
+	if !ok {
+		return nil
+	}
+	topic.s.Cancel()
+	topic.h.Cancel()
+	if err := m.ps.UnregisterTopicValidator(id.String()); err != nil {
+		return err
+	}
+	if err := topic.t.Close(); err != nil {
+		return err
+	}
+	delete(m.m, id)
+	return nil
+}
+
+func (m *TopicManager) topicValidator(context.Context, peer.ID, *pubsub.Message) bool {
+	// @todo: determine if this is needed (related to host signatures)
+	return true
+}
+
+// Publish a record request to a thread.
+func (m *TopicManager) Publish(ctx context.Context, id thread.ID, req *pb.PushRecordRequest) error {
+	m.RLock()
+	defer m.RUnlock()
+	topic, ok := m.m[id]
+	if !ok {
+		return fmt.Errorf("thread topic not found")
+	}
+
+	data, err := req.Marshal()
+	if err != nil {
+		return err
+	}
+	return topic.t.Publish(ctx, data)
+}
+
+// watch peer events from a pubsub topic.
+func (m *TopicManager) watch(ctx context.Context, id thread.ID, topic *topic) {
+	for {
+		pe, err := topic.h.NextPeerEvent(ctx)
+		if err != nil {
+			break
+		}
+		var msg string
+		switch pe.Type {
+		case pubsub.PeerJoin:
+			msg = "JOINED"
+		case pubsub.PeerLeave:
+			msg = "LEFT"
+		}
+		log.Infof("pubsub peer event: %s %s %s", pe.Peer.String(), msg, id.String())
+	}
+}
+
+// subscribe to a topic for thread updates.
+func (m *TopicManager) subscribe(ctx context.Context, id thread.ID, topic *topic) {
+	var err error
+	topic.s, err = topic.t.Subscribe()
+	if err != nil {
+		log.Errorf("error subscribing to topic %s: %s", id.String(), err)
+		return
+	}
+
+	for {
+		msg, err := topic.s.Next(ctx)
+		if err != nil {
+			break
+		}
+		from, err := peer.IDFromBytes(msg.From)
+		if err != nil {
+			log.Errorf("error decoding publisher ID: %s", err)
+			continue
+		}
+		if from.String() == m.host.String() {
+			continue
+		}
+
+		req := new(pb.PushRecordRequest)
+		if err = proto.Unmarshal(msg.Data, req); err != nil {
+			log.Errorf("error unmarshaling published record", err)
+			continue
+		}
+
+		log.Debugf("received multicast record from %s", from.String())
+
+		m.handle(req)
+	}
+}
```

Comment:
Some comments/questions:

1) The `New..` receives a `ctx` that it's used to cancel new items processing. But the process has no way of knowing if some things is being processed before, say, finishing the server. That might make a hard-stop of the logic at any point of the event handler. Maybe that's safe because all logic have store-transactionality.

2) On every registration, a new goroutine is fired, but I'm not sure they've terminated accordingly on a `Remove` call. Goroutines only depend on the manager's context. So it seems they'd leak.

3) What do you think of making `handler` also receive a `ctx`? Seeing `net/server.go` I see that indirectly uses `n.ctx` but I'm not sure that might be the best. You can use the manager `ctx` in the handler, so to also cancel on-going pubsub handling if the manager receives a cancellation.

Some fuzzy writing can get into more details/suggestions or comments if something isn't clear enough!

Reply:
Regarding 2., I'd ask: that `break` if `.Next` is returned is assumed that the topic was closed? If that's the only case that could error, I guess that's the goroutine finalization logic!

Reply:
1) Right, I think it's safe since records coming in over pubsub end up just hitting the server's `PushRecord` method, as if they had come in _normally_, which has its own safety measures.

2) The subscription `.Next` does receive an error "subscription cancelled by calling sub.Cancel()", which triggers the `break`, which is all triggered by `Remove`ing the topic. However, the `EventHandler` does not seems to function that way. I'll see about using a different context.

3) 👍